src/modules/robot/Robot.cpp

```diff
@@ -1289,73 +1346,58 @@ bool Robot::append_milestone(const float target[], float rate_mm_s)
+    float secs = distance / rate_mm_s;
 
-    // check per-actuator speed limits
+    // check per-actuator speed limits by looking at the minimum time each axis can move its specified amount
+    // this is regardless of whether it is mm/sec or deg/sec for a rotary axis
     for (size_t actuator = 0; actuator < n_motors; actuator++) {
+        // actual distance moved for this actuator
+        // NOTE for a rotary axis this will be degrees turned not distance
         float d = fabsf(actuator_pos[actuator] - actuators[actuator]->get_last_milestone());
-        if(d == 0 || !actuators[actuator]->is_selected()) continue; // no movement for this actuator
-
-        float actuator_rate= d * isecs;
-        if (actuator_rate > actuators[actuator]->get_max_rate()) {
-            rate_mm_s *= (actuators[actuator]->get_max_rate() / actuator_rate);
-            isecs = rate_mm_s / distance;
+        if(d < 0.00001F || !actuators[actuator]->is_selected()) continue; // no perceptible movement for this actuator
+
+        // find approximate min time this axis needs to move its distance
+        float actuator_min_time= d / actuators[actuator]->get_max_rate();
+        if (secs < actuator_min_time ) {
+            // this move at the default rate will move too quickly for this actuator
+            // so we decrease the overall feed rate so it can complete within the min time for this actuator
+            rate_mm_s= distance / actuator_min_time;
+            DEBUG_PRINTF("new rate: %f - %d\n", rate_mm_s, actuator);
+            // recalculate time from new rate
+            secs = distance / rate_mm_s;
         }
 
-        // adjust acceleration to lowest found, for now just primary axis unless it is an auxiliary move
-        // TODO we may need to do all of them, check E won't limit XYZ.. it does on long E moves, but not checking it could exceed the E acceleration.
-        if(auxilliary_move || actuator < N_PRIMARY_AXIS) {
-            float ma =  actuators[actuator]->get_acceleration(); // in mm/sec²
-            if(!isnan(ma)) {  // if axis does not have acceleration set then it uses the default_acceleration
-                float ca = fabsf((d/distance) * acceleration);
-                if (ca > ma) {
-                    acceleration *= ( ma / ca );
-                }
+        DEBUG_PRINTF("act: %d, d: %f, distance: %f, min: %f, rate: %f, secs: %f, acc: %f\n", actuator, d, distance, actuator_min_time, rate_mm_s, secs, acceleration);
+
+        // adjust acceleration to not exceed the actuators acceleration
+        float ma =  actuators[actuator]->get_acceleration(); // in mm/sec²
+        if(!isnan(ma)) {  // if axis does not have acceleration set then it uses the default_acceleration
+            float ca = fabsf((d/distance) * acceleration);
+            // if it exceeds the axis acceleration then we reduce the acceleration by the ratio it is over
+            if (ca > ma) {
+                acceleration *= ( ma / ca );
```

Comment:
While mathematically correct, this is numerically not very robust, as you are daisy-chaining multiplications.  You could just take 
`acceleration = ma*distance/d;`
In my solution I use a factor to save repeated `distance/d` calculations.